### DISCUSSION TITLE
Full cache miss when editing a single file for a local project, pnpm, turbo, esbuild

### DISCUSSION THREAD

**Discussion by chrisdeeming** [vercel/turborepo#9179] [Help] [answered]
### Summary

I have probably explained terribly in the title, so here it goes.

I have a local project which allows me to build and bundle our implementation of the Tiptap editor.

In the root of the project I pull in all my dependencies.

I have a `pnpm-workspace.yaml` file:

```yaml
packages:
    - 'packages/*'
```

This is my `turbo.json` file:

```json
{
	"$schema": "https://turbo.build/schema.json",
	"tasks": {
		"build": {
			"inputs": ["src/**/*", "package.json", "esbuild.config.js"],
			"outputs": ["dist/**"]
		}
	}
}
```

This is my `pnpm build` command:

```json
"build": "turbo run build && node esbuild.config.js",
```

This is my `esbuild.config.js`

```
const esbuild = require('esbuild')

const buildOptions = {
	entryPoints: ['./index.js'],
	bundle: true,
	outfile: '../../js/vendor/tiptap/tiptap-bundle.js',
	platform: 'browser',
	target: ['es2020'],
	format: 'iife',
	globalName: 'Tiptap',
	minify: true,
	sourcemap: true,
	define: {
		'process.env.NODE_ENV': '"production"',
	},
}

async function build ()
{
	try
	{
		const result = await esbuild.build(buildOptions)
		console.log('Build completed successfully')

		if (result.errors.length > 0)
		{
			console.error('Build errors:', result.errors)
		}
		if (result.warnings.length > 0)
		{
			console.warn('Build warnings:', result.warnings)
		}

	}
	catch (err)
	{
		console.error('Build failed:', err)
		process.exit(1)
	}
}

build()
```

Within the root directory I also have a `packages` directory which contains a number of extensions I have written for Tiptap to extend the functionality.

Each of those has a `package.json`:

```json
{
  "name": "@me/tiptap-extension-video",
  "version": "1.0.0",
  "description": "video extension for tiptap",
  "private": true,
  "packageManager": "pnpm@9.10.0",
  "main": "dist/index.cjs.js",
  "module": "dist/index.js",
  "types": "dist/index.d.ts",
  "scripts": {
    "clean": "rm -rf dist",
    "build": "node esbuild.config.js"
  },
  "devDependencies": {
    "esbuild": "^0.23.1",
    "typescript": "^5.5.3"
  },
  "peerDependencies": {
    "@tiptap/core": "^2.0.0",
    "@tiptap/pm": "^2.0.0",
    "@me/tiptap-extension-resize-utilities": "workspace:*"
  }
}
```

And its own esbuild config:

```js
const esbuild = require('esbuild')

const baseConfig = {
	entryPoints: ['src/index.ts'],
	bundle: true,
	sourcemap: true,
	minify: process.env.NODE_ENV === 'production',
	target: ['es2020'],
	external: ['@tiptap/core', '@tiptap/pm', '@me/tiptap-extension-resize-utilities'],
}

const cjsConfig = {
	...baseConfig,
	outfile: 'dist/index.cjs.js',
	format: 'cjs',
}

const esmConfig = {
	...baseConfig,
	outfile: 'dist/index.js',
	format: 'esm',
}

async function build ()
{
	try
	{
		await Promise.all([
			esbuild.build(cjsConfig),
			esbuild.build(esmConfig),
		])
		console.log('Build completed successfully')
	}
	catch (error)
	{
		console.error('Build failed:', error)
		process.exit(1)
	}
}

build()
```

I apologise in advance for how detailed this is but I figured if there was an issue it would be better to get everything out there in one go.

Now, everything is working fine. I really enjoy Turbo over Lerna, for example, and even with this issue it seems to be super fast.

But the specific issue I'm facing is that if I modify one `.ts` file in one of the packages, the next time I build, every single package is a cache miss. I have 18 packages in total. It will report there are 0 cached:

```
Cached:    0 cached, 18 total
```

I subsequent run without changes, as expected, will be fully cached:

```
Cached:    18 cached, 18 total
```

If I edit one file in one package and build my expectation was it would look something like this:

```
Cached:    17 cached, 18 total
```

Are my expectations faulty in some way or is there an issue here? Happy to provide any log output (or look for specifics in output) that might help narrow this down.

Thank you!

**Comment by anthonyshew** [comment, marked as answer]
> In the root of the project I pull in all my dependencies.

Does this mean you've installed dependencies into the root `package.json`? Like:
```
{
  "dependencies": {
    "@repo/config": "workspace:*"
  }
}
```

If so, then any change to `@repo/config` will invalidate all packages, since changes to root dependencies affect all packages.

**Reply by chrisdeeming** [reply]
Well, now I feel very silly 🙂

So yeah, that's exactly what I was doing. My entry point was referencing the packages hence them being listed as dependencies, not realising that I could just import them via the path to the `dist` directory for each package.

With that resolved, I can now modify code in a package and get a much more reasonable build:

```
 Tasks:    18 successful, 18 total
Cached:    17 cached, 18 total
  Time:    468ms 
```

Thank you for your help with this. I did feel that there was something off about my approach so thanks for helping me see the light.
